doc/rm/asm_coding_style.md

```diff
@@ -0,0 +1,275 @@
+ */
+compute_stuff:
+  .globl compute_stuff
+  li a0, 0xdeadbeef
+  li a1, 0xbeefcafe
+  li a2, 0xcafedead
+  ret
+```
+
+### Ending an Instruction Sequence
+
+***Every code path within an assembly file must end in a non-linking jump.***
+
+Assembly should be written such that the program counter can't wander off past the written instructions.
+As such, all assembly should be ended with `ret` (or any of the protection ring returns like `mret`), an infinite `wfi` loop, or an instruction that is guaranteed to trap and not return, like an `exit`-like syscall.
+
+Example:
+```S
+loop_forever:
+  wfi
+  j loop_forever
+```
+
+### Alignment Directives
+
+***Do not use `.align`; use `.p2align` and `.balign` as the situation requires.***
+
+The exact meaning of `.align` depends on architecture; rather than asking readers to second-guess themselves, use alignment directives with strongly-typed arguments.
+
+Example:
+```S
+  // Correct:
+  .balign 8 // 8-byte aligned.
+  tail _magic_symbol
+
+  // Wrong:
+  .align 8 // Is this 8-byte aligned, or 256-byte aligned?
+  tail _magic_symbol
+```
+
+### Inline Binary Directives
+
+***Always use `.byte`/`.2byte`/`.4byte`/`.8byte` for inline binary data.***
+
+`.word`, `.long`, and friends are confusing, for the same reason `.align` is.
+
+## Inline Assembly
+
+This advice applies to function-scope inline assembly in `.c` and `.cc` files.
+For an introduction on this syntax, check out [GCC's documentation](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html).
+
+### When to Use
+
+***Avoid inline assembly as much as possible, as long as correctness and readability are not impacted.***
+
+Inline assembly is best reserved for when a high-level language cannot express what we need to do, such as expressing complex control flow or talking to the hardware.
+If a compiler intrinsic can achieve the same effect, such as `__builtin_clz()`, then that should be used instead.
+
+> The compiler is *always* smarter than you; only in the rare case where it is not, assembly should be used instead.
+
+### Formatting
+
+Inline assembly statements must conform to the following formatting requirements, which are chosen to closely resemble how Google's clang-format rules format function calls.
+- Neither the `asm` or `__asm__` keyword is specified in C; the former must be used, and should be `#define`d into existence if not supported by the compiler.
+  C++ specifies `asm` to be part of the grammar, and should be used exclusively.
+- There should not be a space after the `asm` qualfiers and the opening parentheses:
+  ```c
+  asm(...);
+  asm volatile(...);
+  ```
+- Single-instruction `asm` statements should be written on one line, if possible:
+  ```c
+  asm volatile("wfi");
+  ```
+- Multiple-instruction `asm` statements should be written with one instruction per line, formatted as follows:
+  ```c
+  asm volatile(
+    "  la sp, _stack_start;"
+    "  tail _crt0;         "
```

Comment:
What's the purpose of having that many spaces before the closing `"`? Seems to make editing harder for no real readability gain IMHO.

Reply:
I'll get rid of it if other people don't like it. Specifically, I'm just imitating the standard multi-line macro style:
```c
#define MY_MACRO \
  int x = 0;                   \
  return x;
```

Reply:
Yeah, let's remove trailing whitespace from assembly string literals. Leading whitespace is useful and should be kept.

Reply:
Fine, fine. I've also added an example label, and formatted the instruction operands to start on the same column.